Add --chown flag to ADD/COPY commands

estesp opened the pull request:
Carry #28499 (which was a carry of #27303)

This adds `--chown` function to the `ADD` and `COPY` commands in `Dockerfile`. The format of the `--chown` flag allows any uid/username and gid/groupname combination separated by a colon, e.g.:
```
--chown=someuser:555
--chown=anyuser:anygroup
--chown=1001:1002
--chown=333:agroupname
```

Lookups (to translate name -> ID integer) will use the `/etc/passwd` and `/etc/group` files found in the container filesystem. Currently if they do not exist, then the `docker build` will fail. This PR probably needs a way to fail gracefully (or ignore failures) on Windows as those containers will not have these files. If only numeric IDs are used, no lookups will be performed.

User namespaces are supported and IDs will be shifted (after translated names to integers) based on the user namespace mappings set in the daemon.

I also think a few more tests are necessary to fully test the behavior, but wanted to get the rebased code available for review initially.

ardnaxelarak committed:
Add --chown flag to Dockerfile ADD and COPY

Rebased by @estesp

Signed-off-by: Kara Alexandra <kalexandra@us.ibm.com>
Signed-off-by: Phil Estes <estesp@linux.vnet.ibm.com>

estesp committed:
Rebase --chown function for ADD/COPY

Rebases and completes initial PR for (prior: --user) --chown flag for
ADD/COPY commands in Dockerfile.

Docker-DCO-1.1-Signed-off-by: Phil Estes <estesp@linux.vnet.ibm.com>

shouze commented on builder/dockerfile/copy.go:
```diff
@@ -56,6 +56,7 @@ type copyInstruction struct {
 	cmdName                 string
 	infos                   []copyInfo
 	dest                    string
+	flChown                 string
```
does `fl` stands for flag?

estesp replied:
yes, used commonly in the codebase; although in this struct the flag value is being stored, so toss up whether I should be using the `fl` prefix at this point :)

dnephin replied:
Ya, maybe this should  be `user` or something like that?

shouze replied:
@dnephin it's a string serialized `uid:gid` id pair, maybe it worth to parse the chown option earluer & split into uid & gid in the struct in place of a single string?

shouze replied:
If yes could be named `destUid` & `destGid`? Or simply `uid` & `gid` of course.

dnephin replied:
We have an object for that, `idtools.IDPair`

dnephin replied:
`s/flChown/idPair/` or something like that. At this point I don't think it's relevant that this came from a flag.

CWSpear commented:
This was quite a journey. I Googled my issue (the whole "having to chown" stuff) and found an issue from 2 years ago.

Then I had to keep reading through massive threads, with both Docker contributors and the community supporting this, but kept having to pick out "continuing this thread [here]," and about 7 or 8 threads later, I find this, which is obviously a very recent PR.

So... where are we with all this? :)

estesp commented:
Hopefully as close as it has ever been to reality! With the maintainers discussion and comments (like [this one](https://github.com/moby/moby/issues/30110#issuecomment-316790885) from issue #30110 I would expect it is just a matter of review and cleaning up a few things with my implementation.

Given @justincormack wants it, I assume he will be a huge supporter and reviewer of my PR, right Justin! =)

justincormack commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +171,53 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+		return false
+	})
+	if err != nil {
+		return -1, err
```
Minor thing, but it is kind of odd returning `-1` here, as `-1` in chown specifically means "do not chown this part" eg so you can chown a user but not a group, but you are not actually using that, so returning `0` ie the empty `int` is probably clearer. I had to double check you were not using the value regardless of the error code...

estesp replied:
Sure; was somewhat of a weird paranoia considering some future caller assuming that the translation was a valid uid/gid (0) by ignoring the `err`, but I agree that there is no actual valid reason to pass `-1` given the `err`.

justincormack commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -602,6 +602,34 @@ RUN [ $(cat "/test dir/test_file6") = 'test6' ]`, command, command, command, com
+RUN echo 'test2:x:1002:' >> /etc/group
+ADD --chown=test1:1002 . /new_dir
+RUN ls -l /
+RUN [ $(ls -l / | grep new_dir | awk '{print $3":"$4}') = 'test1:test2' ]`)),
```
I would be inclined to add a test that does `ls -ln` to chec the numberi values too as well as the names...

estesp replied:
sounds good

justincormack commented:
Some minor comments but otherwise looks good. I think it is ok to fail on named users on Windows for now, maybe have a test for platform rather than trying to look for `/etc/passwd`.

dnephin commented on builder/dockerfile/copy.go:
```diff
@@ -405,26 +407,35 @@ func performCopyForInfo(dest copyInfo, source copyInfo, options copyFileOptions)
+		// to the requested --chown flags as provided to the original command
+		if err := os.Lchown(dest, chownPair.UID, chownPair.GID); err != nil {
+			return errors.Wrapf(err, "failed to chown target directory")
+		}
```
This (and the `isExistingDirectory()` call above) is unnnecessary. It's already handled by `fixPermissions()`

estesp replied:
except that it doesn't work for the chown case. I took the longest time trying to decide it if was worth adding complexity to `fixPermissions()` or not. The case that doesn't work is when the new directory is actually created by `archive.CopyWithTar`, so when checked for existence in `fixPermissions`, it *does* exist, so the wrong codepath is taken for the chown case. What I wanted was to be able to make an archiver with `ChownOpts` just for this `ADD` or `COPY` command, but after spending time in that code, it isn't possible without significant changes to add `TarOptions` cascading through a bunch of functions that don't take it today.

dnephin replied:
Ah that makes sense, which means this is probably a bug introduced when this was moved out of daemon.

I think the logic should still be in `fixPermission()`. We can pass in the argument as a param, which is what the old code did I believe.

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -157,6 +158,7 @@ func add(req dispatchRequest) error {
 	if err != nil {
 		return err
 	}
+	copyInstruction.flChown = flChown.Value
```
Are we sure we want to add this to `ADD` as well? We only have `--from` on `COPY`, and I think that's by design.

I thought `ADD` was kind of being deprecated/frozen.

estesp replied:
Both prior PRs were adding the same feature to both so I simply followed suit. I didn't see anything in recent discussions narrowing the focus, but I'm willing to hear arguments either way.

thaJeztah replied:
I think we should add it to both (even though we don't want to make `ADD` more magical); if we don't add it to `ADD`, people using it for adding remote files won't be helped by this PR.

dnephin replied:
Shouldn't people using `ADD` for remote files be converting things to `RUN curl` + `COPY` anyway? So by not adding it to `ADD` we're helping nudge people in the right direction.  It doesn't really make sense to have two almost identical instructions.

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -187,6 +190,7 @@ func dispatchCopy(req dispatchRequest) error {
 	if err != nil {
 		return err
 	}
+	copyInstruction.flChown = flChown.Value
```
I think we should parse the Value here so we can fail fast when it's wrong. That way we can store an `IDPair` in `copyInstruction` as well.

estesp replied:
we have to be at a spot where we have the mount information for the container fs.. hence where it happens in this PR. Otherwise there is no source of truth to translate name -> IDs. I can look again and make sure I do that translation as early as we have that info.

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -107,10 +111,17 @@ func (b *Builder) exportImage(state *dispatchState, imageMount *imageMount, runC
+	if inst.flChown != "" {
+		commentStr = fmt.Sprintf("%s --chown=%s %s in %s ", inst.cmdName, inst.flChown, srcHash, inst.dest)
+	} else {
+		commentStr = fmt.Sprintf("%s %s in %s ", inst.cmdName, srcHash, inst.dest)
```
minor nit, these lines are mostly the same so this could just be a variable for `chownComment` and use `%s %s%s %s in %s` as the format.

Could even be a separate function to keep it from cluttering the logic in `performCopy()`

estesp replied:
I should have known this comment was coming! :) Even though I factored it out further per your comments on prior PRs for this, there is more trimming possible for sure :)

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -125,9 +136,32 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+		if err != nil {
+			return errors.Wrapf(err, "unable to convert uid/gid to host mapping")
+		}
+	}
```
Please extract this whole block (parsing the uid/gid pair) into a new function, and maybe call that function from `dispatchCopy()` instead of here.

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +171,53 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+	return parts[0], parts[1]
+}
+
+func lookupUser(userStr, filepath string) (int, error) {
```
I think this already exists as `idtools.LookupUser()` / `LookupUID()`? Can we re-use either of those?

estesp replied:
No, per my prior comment

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +171,53 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+	return users[0].Uid, nil
+}
+
+func lookupGroup(groupStr, filepath string) (int, error) {
```
I think this already exists as `idtools.LookupGroup()` / `LookupGID()`? Can we re-use either of those?

estesp replied:
No

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +171,53 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
 	return b.exportImage(state, imageMount, runConfigWithCommentCmd)
 }
 
+func parseChownFlag(flChown string) (string, string) {
```
I'm surprised this doesn't exist in `pkg/idtools`. I think it should probably move to `idtools.ParseIDPair(string) (IDPair, error)`

estesp replied:
because the context is quite different and requires different *knowledge* in each situation where it is used. `pkg/idtools` only considers the case where the host `/etc/{passwd,group}` files are used for translation, as well as possible use of `getent`. That would be totally invalid for this case, where we need the **container** files to be used.

dnephin replied:
Could we refactor it further to share more of the code?

shouze replied:
would need to introduce some context yup, but at the price of complexity?

dnephin replied:
I don't think the complexity changes much, it's just being moved somewhere more appropriate, so there's less duplicate code to contain.

dnephin commented:
Looks like windows failures are real because of lchown. I think maybe fixed by removing the code I suggested is not necessary.

shouze commented:
@estesp @dnephin can/will this PR fix #32816?

estesp commented:
@dnephin and others; PR has been updated. A bit of vacation delay, sorry :)

At this point I believe the only review comments not handled is the recommended consolidation of `pkg/idtools` parsing of host `/etc/{passwd,group}` files and the new parsing here of container specific `/etc/` files.

After looking through the `idtools` package use of the `opencontainers/runc/libcontainer/user` package and this new code's use, I don't see any effective way of combining these without simply adding complexity. In essence, the `user` package offers two paths:
 1) the "easy" default path of letting that package read host default files; and
 2) the more complex path of providing that code filepaths or open readers along with a custom filter function to find a specific entry.

The `idtools` package uses the "easy" path, because that fits the need of that package, and this new PR code uses the other path, because we want to reference specific paths inside the container we're using during build. Unless we force `idtools` to take the more complex path, and basically rewrite the "easy path" as a consumer of the functions already in that `user` package. Either way we end up duplicating code by copying functions that currently exist in `user`. I believe the way it is now, we have two consumers of the `user` package from libcontainer; they just use two different layers of that functionality.

g13013 commented:
Any news on that ?

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +159,70 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+		// if no group specified, use the user spec as group as well
+		userStr, grpStr = parts[0], parts[0]
+	}
+	userStr, grpStr = parts[0], parts[1]
```
This line will will panic when `len(parts) == 1` , it's not in an else.

This function should have unit tests.

dnephin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -602,6 +602,36 @@ RUN [ $(cat "/test dir/test_file6") = 'test6' ]`, command, command, command, com
 	}
 }
 
+func (s *DockerSuite) TestBuildChownFlagAdd(c *check.C) {
```
I'm still not convinced we want to add this to `ADD`.

We definitely don't need two slow integration tests for this. The code path is 99% the same. Let's just have a single one for COPY. Even better would be to add it to an existing integration test that does COPY, but I'm not sure if we have one in the API suite yet.

`integration-cli` is deprecated, so these should be API tests. We just merged `integration/` so there isn't a lot of work there yet, but at least this should be in `docker_api_build_test.go`

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +159,70 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+	return chownPair, nil
+}
+
+func lookupUser(userStr, filepath string) (int, error) {
```
This (and lookupGroup) should have unit tests for each case

dnephin commented:
>  I believe the way it is now, we have two consumers of the user package from libcontainer; they just use two different layers of that functionality.

Thanks for looking into it

dnephin commented:
> here is very poor scaffolding for standard use of test files

Not true. All the same scaffolding works for the api tests. See https://github.com/moby/moby/blob/master/integration-cli/docker_api_build_test.go#L288-L306

> I don't see any movement of cli_* content to api_*

There should be. Every PR I have seen has called out to change the cli test to an api test

dnephin commented:
> If you mean only one test given the code paths are very similar, that makes sense to me.

Yes, I think a single test is sufficient.

@tonistiigi Do you think we should add this flag to `ADD` as well? or just `COPY` ?

dnephin commented on builder/dockerfile/internals_test.go:
```diff
@@ -129,3 +132,132 @@ func TestCopyRunConfig(t *testing.T) {
+	} {
+		t.Run(testcase.name, func(t *testing.T) {
+			_, err := parseChownFlag(testcase.chownStr, contextDir, testcase.idMapping)
+			assert.Error(t, err, "Expected error reason: "+testcase.descr)
```
```go
assert.EqualError(t, err, testcase.descr)
```

This is only checking there is any error, it doesn't actually check the error string. With this change the `descr` would change to the expected error string (`No such user: ...`, etc)

dnephin commented on builder/dockerfile/internals_test.go:
```diff
@@ -129,3 +132,132 @@ func TestCopyRunConfig(t *testing.T) {
+			idPair, err := parseChownFlag(testcase.chownStr, contextDir, testcase.idMapping)
+			if err != nil {
+				t.Fatalf("Failed to parse chown flag: %v", err)
+			}
```
You can do `require.NoError(t, err)` which is the preferred way of handling this

dnephin reviewed (commented):
Thanks, unit tests are looking good, just a couple comments

estesp commented:
@dnephin thanks for the review; I've updated the unit tests with your recommendations and moved a single "COPY"-based test into the `docker_api_` build tests from `docker_cli_build_*` and removed the two tests that were there. PTAL.

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +159,80 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+		return 0, err
+	}
+	if len(groups) == 0 {
+		return 0, errors.New("No such group: " + groupStr)
```
Small thing, this will fail golint with the latest version, which we're upgrading to in #34551 . Errors need to start with lowercase.

dnephin commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +159,80 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+		return 0, err
+	}
+	if len(users) == 0 {
+		return 0, errors.New("No such user: " + userStr)
```
same golint problem

dnephin reviewed (approved):
LGTM

but please fix the golint errors

estesp commented:
@mstanleyjones just FYI--this changes user-facing flags for the `ADD` and `COPY` commands in `Dockerfile`; the prior PRs didn't have docs yet, so we need to add them in this PR or a separate docs PR. Any guidance from you on whether you want me to start that or you want to take a look at this and give recommendations?

mstanleyjones commented:
Any flags you add should be self-documenting from the Cobra code. If you want to add any examples that use the flags, add them as level-3 sections under the relevant `## Examples` header in files in https://github.com/docker/cli/tree/master/docs/reference/commandline/.

estesp commented:
Sorry @mstanleyjones, I should have been clearer; these are not client commands, but changes to the accepted commands within a `Dockerfile`; specifically the `ADD` and `COPY` command documentation here: https://docs.docker.com/engine/reference/builder/#add

So, I think this is all manually documented in that section; I don't know of other places there are `Dockerfile` references or examples that might need updating.

mstanleyjones commented:
Aha, I see now! So that content is pulled directly from https://github.com/docker/cli/blob/master/docs/reference/builder.md into the docs.

tonistiigi commented on builder/dockerfile/internals.go:
```diff
@@ -137,6 +159,80 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
+		userStr, grpStr = parts[0], parts[1]
+	}
+
+	uid, err := lookupUser(userStr, filepath.Join(ctrRootPath, "etc", "passwd"))
```
these should use `FollowSymlinkInScope`

vdemeester reviewed (approved):
LGTM 🐸
cc @tonistiigi @estesp

estesp commented:
Thanks @tonistiigi ; updated to use `FollowSymlinkInScope`

tonistiigi commented:
LGTM, userns failures seem unrelated

Silex commented:
In what docker version can we expect this feature to land?

sascha-egerer commented:
> In what docker version can we expect this feature to land?
https://github.com/docker/docker-ce/releases/tag/v17.09.0-ce-rc1

presto9292 commented:
Just tried it on 17.09.0-ce-rc1, the syntax is 
--chown=anyuser:anygroup
rather than
--chown anyuser:anygroup

shouze commented:
@presto I've tested too on the recent 17.09 edge release and effectively requires a `=`.

shouze commented:
@presto9292 I've added a comment in https://github.com/docker/cli/pull/467

uzzzerrrname commented:
Can you add `chmod`, too? I frequently build Linux images from a Windows machine and I'm kind of annoyed with the 755 mode the added/copied files end up with.

rcjsuen commented:
@uzzzerrrname I suggest you open a separate GitHub issue for your `chmod` request instead of asking for it in this PR that has already been merged and closed.

mavericksid commented:
How to use this without losing portability? I'll try to explain with an example wherein I find --chown causing portability issues.

Let's say I got a base image (mybase) which I use for running other containers as non-root user (let's say with UID 1000) using [gosu](https://github.com/tianon/gosu) in ENTRYPOINT.
`exec /usr/local/bin/gosu user "$@"`

I use this base image (mybase) for building other container (let's say mycontainer) Dockerfile where I COPY some files from host to container using --chown
```
FROM mybase
RUN mkdir -p /tmp/build
COPY --chown=1000 . /tmp/build
```

Now I want the files generated via my build tool (from mycontainer) to be shared on a Docker Volume mapped with my host. Generated files would be available at specified host location but the owner of those files would be user with id 1000. Doesn't this mean I'll need to have a user with UID 1000 on host machine for them to be accessible? Isn't there a way to provide value for --chown dynamically? I tired using ENV and ARG but they are not accessible in --chown.

uzzzerrrname commented:
@mavericksid looks like a bug.
in COPY, there is parameter expansion for paths
apparently, it is not the case for `--chown`

mavericksid commented:
Hi @thaJeztah 
Thanks for marking it with correct label. Would you know what should be the suggested workaround for this?

huntsberg commented:
Got the update this AM, and it worked great - it took the build time from "so long that I don't know because I go work on something else" to about a minute.  I've been watching and waiting - thank you for getting this sorted and out!

vincentwoo commented:
I saw this change in the release log for the latest stable version of docker-ce, but it doesn't appear to be in the docs: https://docs.docker.com/engine/reference/builder/#add

Is this actually released?

mwarkentin commented:
@vincentwoo yes, it’s working in docker 17.09-ce.

thaJeztah commented:
See https://github.com/docker/cli/pull/467 for the docs; that PR was a bit delayed 😅

vipseixas commented:
--chown does not work for compressed files (eg: tgz), it seems to only change the owner of the compressed file itself and then ignore the extracted content.

Is it the intended behavior?

thaJeztah commented:
@vipseixas I'd say; yes, because it'll only change permissions on the file you _add_. Archives themselves can have permissions set as well, which are preserved (if I'm not mistaken).

Given that automatically extracting archives only works for local files; can you explain your use-case? 
i.e., instead of compressing local files, adding them to the image, and then decompressing, why not add the uncompressed files directly?

vipseixas commented:
@thaJeztah I have a 3rd party installation (PHP code) that comes as a 180MB tar.gz file and I have to open it inside Apache's /var/www. The uncompressed files grows to ~570MB so it would be easier to store the compressed file with my Dockerfile. But if I ADD the compressed file and use a "RUN chown" I end up with 2 huge layers in my image.

derikson commented:
@vipseixas as a workaround, you could use multi-stage build to extract the files in one build stage, then COPY with chown in the second stage. This would prevent the final image from having the extra layer.

vipseixas commented:
@derikson Your suggestion worked perfectly! I was not aware of this new feature, thanx.

marius311 commented:
This is awesome, thanks guys! Probably one of my most desired features for years. 

At the risk of being one those never-satisfied Docker users though :) ...

@thaJeztah:
> It's not a bug; variable expansion in options is not supported at this moment

Is this planned at all and if so is there an issue we could track this at?

thaJeztah commented:
> Is this planned at all and if so is there an issue we could track this at?

A feature request for that was opened here: https://github.com/moby/moby/issues/35018 (but not decided on yet 😅)

khatribharat commented:
This isn't yet reflected in the [Dockerfile doc](https://docs.docker.com/engine/reference/builder/)

thaJeztah commented:
@khatribharat docs took some time to get merged (see https://github.com/docker/cli/pull/467), but will be updated once Docker 17.12 has been released
